### PR TITLE
State bo

### PR DESCRIPTION
**Pull request by fedeallocati**
Still have to add examples and tests (I'll be adding tests with every acquisition function/stopping criterion/inner optimization fuction, to check that I didn't broke any), but I made the pull request so you can start looking at the code.
I was thinking whether we should specify the default reward function as NoReward everywhere (that means every function/class that receives a  Reward function as argument), or just on the BOptimizer functions. I have a little dilema with this:
At first, it seemed to me that it would be best for the user to have it defaulted everywhere, but then I saw that this functor is always passed from the BOptimizer to the rest of the classes, and that having it as default could introduce difficult to detect bugs (e.g. if we forget to pass the functor somewhere, it would use the default, instead of what the user provided, and there won't be any compilation errors or anything).
Would like to hear your thoughts.
In the next days I will add the examples of usage and tests.
CC: @Aneoshun @costashatz @jbmouret

### CODE REVIEW

**Comment by jbmouret** [comment]
I am wondering whether "RewardFunction" is the right word here. It is very RL/Learning-oriented and this concept may actually be more general. Is there a better word? This could be an "Aggregator" or a "EvalFunction"? 

In addition, the code should also work with a composite GP (n GPs in a class).

**Comment by fedeallocati** [comment]
Totaly agree that RewardFunction isn't the best term, but I couldn't think
of a better one. I'm thinking of "Fitness" but that seems oriented to
evolutionary algorithms. When minimizing functions "Cost" always seems
appropiate. What would be the equivalent when maximizing?

I'm not sure I see a reson why the code would'nt work with composite GP, as
long as the mean is a Vector (wheter taken from a multi dimensional GP or n
unidimensional GPs), it should work ok.

2015-10-18 22:52 GMT+02:00 JB Mouret <notifications@github.com>:

**Comment by fedeallocati** [comment]
As a by product of trying to test that I didn't broke anything with the State Based BO, I made a test that declares every possible combination of acquisition function, gp, mean, kernel, inner optimization and init function plus always using all of the stats and stopping criterion (still in development, requieres lots of copypaste and manual work). With this, lots of small bugs unrelated to StateBO are beginning to appear. I fixed some of them in my last commits, but now I'm wondering if it's not better to make another branch just for this, to not make this pull request enormous.
So what I'm propoisng is that you check this pull request as it is now, see if everything related to StateBO makes sense, and in case it does, approve it. 
Then, I'll make another branch for solving all the small bugs that appear.

**Comment by fedeallocati** [comment]
PD. You can check an example of usage of the StateBO in examples/obs_multi.cpp

**Comment by costashatz** [comment]
@fedeallocati I totally agree with you. Each pull request should be as small as possible so that it is easy to check. I will check the examples/code in more detail and confirm this PR tomorrow.

File: src/examples/obs_multi.cpp [hunk 1]
```diff
@@ -40,53 +37,48 @@ struct Params {
...
+        res(1) = -5 * x(1) + 2;
         return res;
     }
 };
 
+struct Average {
+    typedef double result_type;
+    double operator()(const Eigen::VectorXd& x) const
+    {
+        return (x(0) + x(1)) / 2;
```
Comment (jbmouret) [inline src/examples/obs_multi.cpp]: Minor detail but you should call it "Average2" or use x.sum()/x.length() (probably better)

**Comment by jbmouret** [comment]
I had a look at the current state and it looks fine for me.

With clang, there is an annoying warning in macros.hpp:

```
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:25:22: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
#define __VA_ARG_N( \ 
                     ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:26:51: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
         _1, _2, _3, _4, _5, _6, _7, _8, _9,_10, \ 
                                                  ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:27:51: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        _11,_12,_13,_14,_15,_16,_17,_18,_19,_20, \ 
                                                  ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:28:51: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        _21,_22,_23,_24,_25,_26,_27,_28,_29,_30, \ 
                                                  ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:29:51: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        _31,_32,_33,_34,_35,_36,_37,_38,_39,_40, \ 
                                                  ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:30:51: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        _41,_42,_43,_44,_45,_46,_47,_48,_49,_50, \ 
                                                  ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:31:51: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        _51,_52,_53,_54,_55,_56,_57,_58,_59,_60, \ 
                                                  ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:33:21: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
#define __RSEQ_N() \ 
                    ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:34:26: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        63, 62, 61, 60, \ 
                         ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:35:50: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        59, 58, 57, 56, 55, 54, 53, 52, 51, 50, \ 
                                                 ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:36:50: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        49, 48, 47, 46, 45, 44, 43, 42, 41, 40, \ 
                                                 ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:37:50: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        39, 38, 37, 36, 35, 34, 33, 32, 31, 30, \ 
                                                 ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:38:50: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        29, 28, 27, 26, 25, 24, 23, 22, 21, 20, \ 
                                                 ^
/Users/jbm/Documents/git/limbo/src/limbo/macros.hpp:39:50: warning: backslash and newline separated by space [-Wbackslash-newline-escape]
        19, 18, 17, 16, 15, 14, 13, 12, 11, 10, \ 
```

**Comment by fedeallocati** [comment]
Probably that is what was breaking the macro when formatting. I'm fixing it. Also will compile with gcc to see if there are any warnings.

**Comment by fedeallocati** [comment]
Done both things

**Comment by costashatz** [comment]
**Seems legit to me!**

**Comment by costashatz** [comment]
**I approve this PR**. Everything seems fine!

**Comment by jbmouret** [comment]
**OK for me**

**Pull request merged**